Use @ByteString annotation to toggle encoding/decoding of byte strings

twyatt opened the pull request:
This is a continuation of the Kotlin/kotlinx.serialization#846 effort, whereas introducing a new `SerialKind` was discouraged: https://github.com/Kotlin/kotlinx.serialization/pull/846#issuecomment-643512265

Instead, a dedicated annotation (`@ByteString`) is used to differentiate `ByteArray` encoding/decoding between type 2 and type 4 (details re: [Major types] are listed in RFC 7049).

Diff against upstream `Kotlin/kotlinx.serialization`'s `dev` branch can be viewed [here](https://github.com/Kotlin/kotlinx.serialization/compare/dev...twyatt:CBOR/ByteString?expand=1).

[Major types]: https://tools.ietf.org/html/rfc7049#section-2.1

twyatt committed:
Use @ByteString annotation to toggle encoding/decoding of byte strings

This is a continuation of the Kotlin/kotlinx.serialization#846 effort,
whereas introducing a new `SerialKind` was discouraged:

https://github.com/Kotlin/kotlinx.serialization/pull/846#issuecomment-643512265

Instead, a dedicated annotation (`@ByteString`) is used to differentiate
`ByteArray` encoding/decoding between type 2 and type 4 (details re:
[Major types] are listed in RFC 7049).

[Major types]: https://tools.ietf.org/html/rfc7049#section-2.1

twyatt committed:
Add basic documentation re: @ByteString annotation

Chris-Corea reviewed (approved):
💥

qwwdfsad committed:
[API stabilization] Exceptions rework (#880)


* Make SerializationException descendant of IllegalArgumentException, provide full-blown constructors capabilities
* Hide JsonException family, MissingFieldException and UnknownFieldException from public API
* Remove obsolete UpdateNotSupportedException

qwwdfsad committed:
[API stabilization] Single entry point for serializers retrieval, Serializers.kt

    * Rename serializerByTypeToken to serializer
    * Internal bookkeeping, cleanup, leverage Kotlin 1.4 improvements

cedrickcooke commented on formats/cbor/commonTest/src/kotlinx/serialization/cbor/SampleClasses.kt:
```diff
@@ -31,7 +31,7 @@ data class NumberTypesUmbrella(
-    val foo: ByteString,
-    val bar: ByteString
+data class ByteArrayUmbrella(
+    @ByteString val foo: ByteArray,
+    @ByteString val bar: ByteArray
```
Would it make more sense to have one field of each encoding, instead of both byte strings?

twyatt replied:
Thanks for catching this. I intended to have one of each and switched to both be `@ByteString` because the assertion payload had them as both byte string (and I was trying to make the test pass). I'll fix this.

twyatt committed:
Merge branch 'dev' into CBOR/ByteString

twyatt committed:
Unify String and ByteString implementations

qwwdfsad committed:
[API stabilization] Rework JSON DSL API (#881)

* Make it aligned with Kotlin standard library where receiver of buildList/buildMap is MutableList/MutableMap
* Make it aligned with our own new naming scheme where all JSON subtypes are mentioned explicitly

qwwdfsad committed:
[API stabilization] Document core annotations, increase retention lev… (#883)

Document core annotations, increase retention level to binary to enable potential future work

Co-authored-by: Leonid Startsev <sandwwraith@users.noreply.github.com>

qwwdfsad committed:
[API stabilization] Builtin package stabilization

    * Deprecate redundant extensions
    * Merge all serializers into single file
    * Extract AbstractDecoder and AbstractEncoder to a separate package

twyatt committed:
Merge branch 'dev' into CBOR/ByteString

twyatt committed:
Minor documentation updates

twyatt committed:
Fix decode indefinite test to use indefinite byte string

twyatt commented:
Closing in favor of https://github.com/Kotlin/kotlinx.serialization/pull/898.
